pkg/bloomgateway/multiplexing.go

```diff
@@ -0,0 +1,225 @@
+	key, err := ulid.New(ulid.Now(), nil)
+	if err != nil {
+		return Task{}, nil, nil, err
+	}
+	errCh := make(chan error, 1)
+	resCh := make(chan v1.Output, 1)
+	task := Task{
+		ID:      key,
+		Tenant:  tenantID,
+		Request: req,
+		ErrCh:   errCh,
+		ResCh:   resCh,
+	}
+	return task, resCh, errCh, nil
+}
+
+// Copy returns a copy of the existing task but with a new slice of chunks
+func (t Task) Copy(refs []*logproto.GroupedChunkRefs) Task {
+	return Task{
+		ID:     t.ID,
+		Tenant: t.Tenant,
+		Request: &logproto.FilterChunkRefRequest{
+			From:    t.Request.From,
+			Through: t.Request.Through,
+			Filters: t.Request.Filters,
+			Refs:    refs,
+		},
+		ErrCh: t.ErrCh,
+		ResCh: t.ResCh,
+	}
+}
+
+// Bounds returns the day boundaries of the task
+func (t Task) Bounds() (time.Time, time.Time) {
+	return getDayTime(t.Request.From), getDayTime(t.Request.Through)
+}
+
+func (t Task) ChunkIterForDay(day time.Time) v1.Iterator[*logproto.GroupedChunkRefs] {
+	cf := filterGroupedChunkRefsByDay{day: day}
+	return &FilterIter[*logproto.GroupedChunkRefs]{
+		iter:      v1.NewSliceIter(t.Request.Refs),
+		matches:   cf.contains,
+		transform: cf.filter,
+	}
+}
+
+type filterGroupedChunkRefsByDay struct {
+	day time.Time
+}
+
+func (cf filterGroupedChunkRefsByDay) contains(a *logproto.GroupedChunkRefs) bool {
+	from, through := getFromThrough(a.Refs)
+	if from.Time().After(cf.day.Add(Day)) || through.Time().Before(cf.day) {
+		return false
+	}
+	return true
+}
+
+func (cf filterGroupedChunkRefsByDay) filter(a *logproto.GroupedChunkRefs) *logproto.GroupedChunkRefs {
+	minTs, maxTs := getFromThrough(a.Refs)
+
+	// in most cases, all chunks are within day range
+	if minTs.Time().Compare(cf.day) >= 0 && maxTs.Time().Before(cf.day.Add(Day)) {
+		return a
+	}
+
+	// case where certain chunks are outside of day range
+	// using binary search to get min and max index of chunks that fall into the day range
+	min := sort.Search(len(a.Refs), func(i int) bool {
+		start := a.Refs[i].From.Time()
+		end := a.Refs[i].Through.Time()
+		return start.Compare(cf.day) >= 0 || end.Compare(cf.day) >= 0
+	})
+
+	max := sort.Search(len(a.Refs), func(i int) bool {
+		start := a.Refs[i].From.Time()
+		return start.Compare(cf.day.Add(Day)) > 0
+	})
+
+	return &logproto.GroupedChunkRefs{
+		Tenant:      a.Tenant,
+		Fingerprint: a.Fingerprint,
+		Refs:        a.Refs[min:max],
+	}
+}
+
+type Predicate[T any] func(a T) bool
+type Transform[T any] func(a T) T
+
+type FilterIter[T any] struct {
+	iter      v1.Iterator[T]
+	matches   Predicate[T]
+	transform Transform[T]
+	cache     T
+	zero      T // zero value of the return type of Next()
+}
+
+func (it *FilterIter[T]) Next() bool {
+	next := it.iter.Next()
+	if !next {
+		it.cache = it.zero
+		return false
+	}
+	for next && !it.matches(it.iter.At()) {
+		next = it.iter.Next()
+		if !next {
+			it.cache = it.zero
+			return false
+		}
+	}
+	it.cache = it.transform(it.iter.At())
```

Comment:
I'd like to avoid the `it.transform(it.iter.At())` function call every time `it.At()` is called.
Depending on the transform function, it could be expensive to do so.